Chatkit/Subscriptions/Subscription.swift

```diff
@@ -0,0 +1,301 @@
+                self.state = .notSubscribed
+
+                let error = Self.onEndReceivedWhileSubscribingError
+                
+                self.delegate?.subscription(self, didReceiveError: error)
+                
+                for completion in completions {
+                    completion(.failure(error))
+                }
+                
+            case let .subscribingStageTwo(_, resumableSubscription, completions):
+                
+                resumableSubscription.onOpen = nil
+                resumableSubscription.onError = nil
+                resumableSubscription.end()
+                
+                // The ORDER of the code here is VITAL:
+                //   We MUST set the state before we call the delegate/completions
+                //   We MUST invoke the delegate method before the completions
+                
+                self.state = .notSubscribed
+                
+                let error = Self.onEndReceivedWhileSubscribingError
+                
+                self.delegate?.subscription(self, didReceiveError: error)
+                
+                for completion in completions {
+                    completion(.failure(error))
+                }
+                
+            case .subscribed:
+                
+                self.state = .notSubscribed
+                
+                let error = Self.onEndReceivedWhileSubscribedError
+                
+                self.delegate?.subscription(self, didReceiveError: error)
+                
+            }
+            
+        }
+        
+        // The ORDER of the code here is VITAL:
+        
+        state = .subscribingStageOne(instanceWrapper: instanceWrapper, completions: [completion])
+        
+        let resumableSubscription = instanceWrapper.subscribeWithResume(using: requestOptions,
+                                                                        onOpening: nil,
+                                                                        onOpen: nil,
+                                                                        onResuming: nil,
+                                                                        onEvent: onEvent,
+                                                                        onEnd: onEnd,
+                                                                        onError: onError)
+        
+        state = .subscribingStageTwo(instanceWrapper: instanceWrapper,
+                                     resumableSubscription: resumableSubscription,
+                                     completions: [completion])
+        
+    }
+    
+    private func makeRequestPath(for subscriptionType: SubscriptionType) -> String {
+        switch subscriptionType {
+        case .user:
+            return "/users"
+        case let .room(roomIdentifier):
+            return "/room/\(roomIdentifier)"
+        }
+    }
+    
+    private static var unsubscribeCalledWhileSubscribingError: Error
```

Comment:
We should have a typed error (`enum`) here.

Reply:
Now sorted with a dedicated `SubscriptionError`